### DISCUSSION TITLE
Error Index: Inferred type is XXX, expected YYY

### DISCUSSION THREAD

**Discussion by Shoeboxam** [opendp/opendp#298] [Error Index]
By the time you invoke a function or call a relation, OpenDP has already determined the type of the data you are supposed to pass.
To ensure type safety, OpenDP checks that the types of the arguments you pass are similar to the type the function is expecting.

Fixing this may be as simple as passing integers instead of floats, or it may require additional preprocessing.
One common cause of this error is passing a scalar `epsilon` to a relation that expects an (`epsilon`, `delta`) tuple.

| rust type | description       |
|-----------|---------------------|
| f64       | 64-bit double floating-point number |
| i32       | 32-bit signed integer      |
| u32       | 32-bit unsigned integer      |
| bool      | boolean             |
| String    | UTF-8–encoded text  |
| usize     | integer index       |

Feel free to comment if you're stuck on this error.

**Comment by a-athanasopoulos** [comment]
is there any more efficient way to iterate over a vector than the following?
private_release  = [ my_mesurement(datum) for datum in datas]

**Reply by Shoeboxam** [reply]
Hmm, that code snip suggests to me that `my_measurement` is randomized response. Is that so?

**Comment by a-athanasopoulos** [comment]
I want to apply the laplacian mechanism to many different inputs to
demonstrate the local privacy model.

The problem is that the dataset is large so i am wondering if there is any
vectorised implementation to speed up the process.

**Reply by Shoeboxam** [reply]
Ah, try this:
```python
import opendp.prelude as dp
dp.enable_features("contrib")

m_lap = dp.m.make_laplace(
    input_domain=dp.vector_domain(dp.atom_domain(T=float)),
	input_metric=dp.l1_distance(T=float),
    scale=1.)

print(m_lap([0.] * 100))
```

See docs here: https://docs.opendp.org/en/stable/api/user-guide/measurements/additive-noise-mechanisms.html

**Comment by lo2aayy** [comment]
I'm using AIM from smartnoise and I'm getting this error, I don't understand much what is the error or how I can fix it, the categorical_cols data are labelencoded data and numerical_cols are currently empty:

`Exception: Unexpected error happened during training.
predicate always fails. Error at center: inferred type is i32, expected f64. See https://github.com/opendp/opendp/discussions/298
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/models/aim.py:38, in Aim.train_synthesizer(self)
     37 try:
---> 38     self.synth.fit(self.X_train, preprocessor_eps=self.preprocessor_eps,
     39                    categorical_columns=self.categorical_cols, continuous_columns=self.numerical_cols)
     40 except Exception as e:

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.9/site-packages/snsynth/aim/aim.py:134, in AIMSynthesizer.fit(self, data, transformer, categorical_columns, ordinal_columns, continuous_columns, preprocessor_eps, nullable, *ignore)
    132 self.num_rows = len(data)
--> 134 self.rho = 0 if self.delta == 0 else cdp_rho(self.epsilon, self.delta)
    136 data = pd.DataFrame(train_data, columns=colnames)

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.9/site-packages/snsynth/utils.py:53, in cdp_rho(epsilon, delta)
     52     return make_fix_delta(adp, delta=budget[1])
---> 53 scale = binary_search_param(
     54     make_fixed_approxDP_gaussian,
     55     d_in=1.0, d_out=budget, T=float)
     56 return make_base_gaussian(scale).map(1.)

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.9/site-packages/opendp/mod.py:465, in binary_search_param(make_chain, d_in, d_out, bounds, T)
    414 """Useful to solve for the ideal constructor argument.
    415 
    416 Optimizes a parameterized chain `make_chain` within float or integer `bounds`,
   (...)
    463 1498
    464 """
--> 465 return binary_search(lambda param: make_chain(param).check(d_in, d_out), bounds, T)

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.9/site-packages/opendp/mod.py:529, in binary_search(predicate, bounds, T, return_sign)
    528 if bounds is None:
--> 529     bounds = exponential_bounds_search(predicate, T)
    531 if bounds is None:

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.9/site-packages/opendp/mod.py:662, in exponential_bounds_search(predicate, T)
    660         error = f". Error at center: {err}"
--> 662     raise ValueError(f"predicate always fails{error}")
    664 center, sign = binary_search(exception_predicate, bounds=exception_bounds, T=T, return_sign=True)

ValueError: predicate always fails. Error at center: inferred type is i32, expected f64. See https://github.com/opendp/opendp/discussions/298

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
File <command--1>:8
      5 del sys
      7 with open(filename, "rb") as f:
----> 8   exec(compile(f.read(), filename, 'exec'))

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/components/synthesizer_trainer.py:123
    120     synthesizer.run_synthesizer()
    122 if __name__ == '__main__':
--> 123     main()

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/components/synthesizer_trainer.py:120, in main()
    118 def main():
    119     synthesizer = SynthesizerTrainer()
--> 120     synthesizer.run_synthesizer()

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/components/synthesizer_trainer.py:111, in SynthesizerTrainer.run_synthesizer(self)
    109 self.split_data()
    110 logger.info('Starting training, this may take some time ..')
--> 111 self.train_model()
    112 logger.info('Finished training')
    113 synth_df = self.synthesize_data()

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/components/synthesizer_trainer.py:76, in SynthesizerTrainer.train_model(self)
     73 else:
     74     model = Quail(self.train_df, self.epsilon, self.metadata['preprocessor_eps'], categorical_cols, numerical_cols, self.target)
---> 76 self.synth = model.fit()

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/models/aim.py:51, in Aim.fit(self)
     48 self.init_synthesizer()
     49 logger.info(f'Synthesizer initialized')
---> 51 self.train_synthesizer()
     52 logger.info(f'AIM synthesizer trained with epsilon of {self.epsilon}')
     54 return self.synth

File /Workspace/Repos/.internal/005c97e368_commits/0c102809b83d85d77faf153c3a78affd04f2433b/src/models/aim.py:42, in Aim.train_synthesizer(self)
     40 except Exception as e:
     41     logger.error(f'Unexpected error happened during training. \n {e}')
---> 42     raise Exception(f'Unexpected error happened during training. \n {e}')

Exception: Unexpected error happened during training. 
 predicate always fails. Error at center: inferred type is i32, expected f64. See https://github.com/opendp/opendp/discussions/298`

**Reply by Shoeboxam** [reply]
Kind of hard to tell without seeing a code example to help reproduce the issue, but can you try setting privacy-loss parameters as floats?

**Comment by yapatta** [comment]
Great thank you for such a great library!!
By the way, I have encountered the issue like this when transforming List[List[str]] in python to a dataframe. I am wondering this issue is related to Rust. Could you kindly tell me how to solve this issue?

```py
    trans_example = make_create_dataframe(
        ["test"]
    ) >> make_select_column("test", str)
    l: List[List[str]] = [['37631915'], ['98762579'], ['83098666'], ['9970619'], ['75973458'], ['4345911'], ['74953876'], ['92845191']]
    trans_example(l)
```
/venv/lib/python3.8/site-packages/opendp/typing.py", line 375, in assert_is_similar
    raise TypeError(f"inferred type is {inferred.origin}, expected {expected.origin}. See {ERROR_URL_298}")
TypeError: inferred type is Vec, expected alloc::vec::Vec. See https://github.com/opendp/opendp/discussions/298
